### PR TITLE
196 - add notification service

### PR DESCRIPTION
**Pull request by Its-OP**
**EventBus**
- Persistent connection was added. It handles reconnections, abrupt disconnects, etc.
- In-memory subscription manager was added. It handles subscriptions of specific handlers to specific events.
- Entities for event bus were added, such as IntegrationEvent and EventHandler.
- Registrations file was created in order to make it easier to add all the EventBus-related infrastructure to a project.

**Notifications**
- NotificationsHub was added. It utilizes SignalR infrastructure to push data to the client.
- Services for each type of notifications were added. For now, they are only needed to return all the notifications, that were sent to a specific user.
- Controlles for each type of notification services were added.
- Entities for each type of notifications were added.

**Overall project**
- Some services and controllers were expanded, so they can perform actions that send notifications.
- EventBus registrations and EventBus configs were added to the Core project.

### CODE REVIEW

**Comment by Its-OP** [comment]
Diagram of the functionality that I'm working on

Notifications

**Review by Vadis00 — DISMISSED** [review]

File: backend/Tasque.Core.WebAPI/Controllers/TaskController.cs [hunk 1]
```diff
@@ -75,5 +75,12 @@ public async Task<IActionResult> GetTasksType()
             return Ok();
             //   return Ok(await _taskService.GetTasksType());
```
Comment (ismv) [inline backend/Tasque.Core.WebAPI/Controllers/TaskController.cs:78]: do we need this commented part?

File: backend/Tasque.Core.Common/Entities/SubscriptionInfo.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Tasque.Core.Common.Entities
+{
+    public class SubscriptionInfo
+    {
+        public Type HandlerType { get; }
+        public SubscriptionInfo(Type handlerType)
+        {
+            HandlerType=handlerType;
```
Comment (ismv) [inline backend/Tasque.Core.Common/Entities/SubscriptionInfo.cs]: spacing

File: backend/Tasque.BLL/Services/TaskService.cs [hunk 3]
```diff
@@ -217,6 +226,32 @@ private int UpdateProjectCounter(int projectId)
...
+        {
+            var task = _dbContext.Tasks.Single(t => t.Id == dto.TaskId);
+            var comment = new Comment
+            {
+                Message = dto.Message,
+                AuthorId = dto.AuthorId,
+                TaskId = dto.TaskId
+            };
+            _dbContext.Comments.Add(comment);
+            //task.Comments.Add(dto.Comment);
```
Comment (ismv) [inline backend/Tasque.BLL/Services/TaskService.cs]: do we need this commented code?

File: backend/Tasque.DAL/Migrations/DataContextModelSnapshot.cs [hunk 4]
```diff
@@ -34,7 +34,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
 
                     b.HasIndex("TasksId");
 
-                    b.ToTable("AttachmentTask", (string)null);
```
Comment (ismv) [inline backend/Tasque.DAL/Migrations/DataContextModelSnapshot.cs:37]: https://github.com/BinaryStudioAcademy/bsa-2022-tasque/pull/225/files#diff-768a9f8593b2787c8140a407517a127d9e1d1d512b13f17621fde11d43856d68

please contact @DaniilPoiarkov regarding this change and find an agreement on reasoning behind this change

File: backend/Tasque.Core.WebAPI/Controllers/TaskController.cs (hunk 1 above)
Comment (Its-OP) [reply backend/Tasque.Core.WebAPI/Controllers/TaskController.cs:78]: We have not implemented the required method yet. I've added TODO to this comment.

File: backend/Tasque.BLL/Services/TaskService.cs (hunk 3 above)
Comment (Its-OP) [reply backend/Tasque.BLL/Services/TaskService.cs]: Resolved

File: backend/Tasque.DAL/Migrations/DataContextModelSnapshot.cs (hunk 4 above)
Comment (Its-OP) [reply backend/Tasque.DAL/Migrations/DataContextModelSnapshot.cs:37]: I've contacted Daniil, he does not think that these `(string)null` parameters pose any value. [Apparently](https://stackoverflow.com/questions/69818798/rolling-back-an-ef-core-migration-does-not-exactly-restore-the-model-snapshot), there is a bug in EF Core 6 where `(string)null` get added to BuildModel on migration remove.

**Review by ismv — APPROVED** [review]

**Review by FeloniukM — APPROVED** [review]

**Pull request merged**